vector/vector.h

```diff
@@ -1,62 +1,325 @@
+std::size_t vector<T>::size() const {
+    return size_;
+}
+
+template<typename T>
+T &vector<T>::front() {
+    return data_[0];
+}
+
+template<typename T>
+T const &vector<T>::front() const {
+    return data_[0];
+}
+
+template<typename T>
+T &vector<T>::back() {
+    return data_[size_ - 1];
+}
+
+template<typename T>
+T const &vector<T>::back() const {
+    return data_[size_ - 1];
+}
+
+template<typename T>
+void vector<T>::push_back(const T &val) {
+    T temp = val;
+    if (size_ != capacity_) {
+        new(data_ + size_) T(val);
+    } else {
+        push_back_realloc(val);
+    }
+    ++size_;
+}
+
+template<typename T>
+void vector<T>::pop_back() {
+    assert(size_ != 0);
+
+    data_[size_ - 1].~T();
+    --size_;
+}
+
+template<typename T>
+std::size_t vector<T>::capacity() const {
+    return capacity_;
+}
+
+template<typename T>
+void vector<T>::reserve(std::size_t val) {
+    if (val > capacity_) {
+        new_buffer(val);
+    }
+}
+
+template<typename T>
+void vector<T>::shrink_to_fit() {
+    if (size_ == capacity_) {
+        return;
+    }
+
+    vector temp = (*this);
+    swap(temp);
+}
+
+template<typename T>
+void vector<T>::clear() {
+    for (int i = 0; i < size_; i++) {
+        data_[i].~T();
+    }
+    size_ = 0;
+}
+
+template<typename T>
+void vector<T>::swap(vector &other) {
+    std::swap(data_, other.data_);
+    std::swap(size_, other.size_);
+    std::swap(capacity_, other.capacity_);
+}
+
+template<typename T>
+typename vector<T>::iterator vector<T>::begin() {
+    return data_;
+}
+
+template<typename T>
+typename vector<T>::iterator vector<T>::end() {
+    return data_ + size_;
+}
+
+template<typename T>
+typename vector<T>::const_iterator vector<T>::begin() const {
+    return data_;
+}
+
+template<typename T>
+typename vector<T>::const_iterator vector<T>::end() const {
+    return data_ + size_;
+}
+
+template<typename T>
+typename vector<T>::iterator vector<T>::insert(vector::const_iterator pos, const T &val) {
+    int index = pos - data_;
+
+    assert(index >= 0 && index <= size_);
+
+    if (capacity_ == 0) {
+        new_buffer(2);
+    } else if (capacity_ == size_) {
+        new_buffer(capacity_ * 2);
+    }
+
+    new(data_ + size_) T();
+    for (int i = size_; i > index; i--) {
+        data_[i] = data_[i - 1];
+    }
+    data_[index] = val;
+    size_++;
+    return data_ + index;
+}
+
+template<typename T>
+typename vector<T>::iterator vector<T>::erase(vector::const_iterator pos) {
+    return erase(pos, pos + 1);
+}
+
+template<typename T>
+typename vector<T>::iterator vector<T>::erase(vector::const_iterator first, vector::const_iterator last) {
+    int from = first - data_;
+    int to = last - data_;
+
+    assert (size_ != 0);
+    assert (0 <= from && from <= size_);   // Не уверен в корректности этой и следующей проверки
+    assert (0 <= to && to <= size_);
+    assert (from <= to);
+
+    int delta = to - from;
+    for (int i = from; i < size_ - delta; i++) {
```

Comment:
[overhead] `size - delta = size - to + from = size - last + data + first - data = size - last + first = size - (last - first)`